spmd/tensor/ops/view_ops.py

```diff
@@ -566,30 +656,66 @@ def reshape_prop(op_schema: OpSchema) -> OutputSharding:
         global_in_shape = input_dtensor_spec.shape
         assert global_in_shape is not None, "Shape required."
 
-        global_out_shape, shard_out = propagate_shape_and_sharding(
+        (
+            global_out_shape,
+            shard_out,
+            shardable_dims,
+        ) = propagate_shape_and_sharding(
             input_dtensor_spec.placements,
             tuple(global_in_shape),
             rules,
             tuple(input_dtensor_spec.mesh.mesh.shape),
         )
-        output_dtensor_spec = DTensorSpec(
-            mesh=input_dtensor_spec.mesh,
-            placements=shard_out,
-            shape=torch.Size(global_out_shape),
-            ndim=len(global_out_shape),
-        )
-        local_out_shape = output_dtensor_spec.local_shape
-
-        # We only need the local shape to lower he call into the local op
-        args = op_schema.args_schema
-        if spec.shape_argnum is not None:
-            op_schema.args_schema = (
-                args[: spec.shape_argnum]
-                + (tuple(local_out_shape),)
-                + args[cast(int, spec.shape_argnum) + 1 :]
+
+        if shard_out is not None:
+            # no reshard needed
+            output_dtensor_spec = DTensorSpec(
+                mesh=input_dtensor_spec.mesh,
+                placements=shard_out,
+                shape=torch.Size(global_out_shape),
+                ndim=len(global_out_shape),
             )
+            local_out_shape = output_dtensor_spec.local_shape
+
+            # We only need the local shape to lower he call into the local op
+            args = op_schema.args_schema
+            if spec.shape_argnum is not None:
+                op_schema.args_schema = (
+                    args[: spec.shape_argnum]
+                    + (tuple(local_out_shape),)
+                    + args[cast(int, spec.shape_argnum) + 1 :]
+                )
+
+            return OutputSharding(output_spec=output_dtensor_spec)
 
-        return OutputSharding(output_spec=output_dtensor_spec)
+        else:
+            # TODO: optimize this. we shouldn't simply blindly replicate
```

Comment:
is this suggesting we might switch sharding dims instead of replicating the unshardable dims?

Reply:
i think would be in general more cost-effective (at least memory-wise), but not sure? didn't think it through.

Reply:
yeah, i think a general reduce_scatter is better and looks similar to zero 3 from memory prospective.